// common-services/dist/log.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.LogOutputTypes = exports.DEFAULT_LOG_ROUTING = exports.DEFAULT_LOG_CONFIG = void 0;

var _knifecycle = require("knifecycle");

let LogOutputTypes;
exports.LogOutputTypes = LogOutputTypes;

(function (LogOutputTypes) {
  LogOutputTypes["OUTPUT"] = "output";
  LogOutputTypes["ERROR"] = "error";
  LogOutputTypes["DEBUG"] = "debug";
})(LogOutputTypes || (exports.LogOutputTypes = LogOutputTypes = {}));

const DEFAULT_LOG_CONFIG = {
  stringify: false
};
exports.DEFAULT_LOG_CONFIG = DEFAULT_LOG_CONFIG;
const DEFAULT_LOG_ROUTING = {
  error: LogOutputTypes.ERROR,
  debug: LogOutputTypes.DEBUG,
  warning: LogOutputTypes.ERROR,
  info: LogOutputTypes.OUTPUT,
  // The stack type allows to filter logs in testing
  // since the stack files paths vary between systems
  // and it is annoying to filter them
  'error-stack': LogOutputTypes.ERROR,
  'debug-stack': LogOutputTypes.DEBUG
};
exports.DEFAULT_LOG_ROUTING = DEFAULT_LOG_ROUTING;

/* Architecture Note #1.1: Logging

I prefer using a unique function with the log type
 in parameter instead of several methods for each
 log types. It is far easyer to mock and to assert
 on logs in my tests.

If provided, I route debug messages to the `debug`
 node module.

*/
var _default = (0, _knifecycle.singleton)((0, _knifecycle.service)(initLog, "log", ["?LOG_CONFIG", "?LOG_ROUTING", "logger"]));
/**
 * Instantiate the logging service
 * @name initLog
 * @function
 * @param  {Object}   services
 * The services to inject
 * @param  {Object}   services.logger
 * The logger object that output the logs
 * @param  {Function} [services.debug = noop]
 * A debugging function
 * @return {Promise<Function>}
 * A promise of the logging function
 * @example
 * import initLog from 'common-services/dist/log';
 *
 * const log = await initLog({
 *   logger: require('winston'),
 *   debug: require('debug')('myapp'),
 *  });
 */


exports.default = _default;

async function initLog({
  LOG_CONFIG = DEFAULT_LOG_CONFIG,
  LOG_ROUTING = DEFAULT_LOG_ROUTING,
  logger
}) {
  log('debug', '👣 - Logging service initialized.');
  return log;
  /**
   * Logging function
   * @param  {String}  type
   * Log type
   * @param  {...*}    args
   * Log contents
   * @return {void}
   * @example
   * log('debug', 'Luke, I am your father!')
   */

  function log(type, ...args) {
    const output = args.map(arg => LOG_CONFIG.stringify ? JSON.stringify(arg) : arg != null ? arg.toString() : '').join(' ');

    if (LOG_ROUTING[type] === LogOutputTypes.ERROR) {
      logger.error(output);
      return;
    }

    if (LOG_ROUTING[type] === LogOutputTypes.OUTPUT) {
      logger.output(output);
      return;
    }

    logger.debug(...args);
  }
}
